genesis/engine/coupler.py

```diff
@@ -44,25 +44,31 @@ def build(self):
 
         if self._rigid_mpm and self.mpm_solver.enable_CPIC:
             # this field stores the geom index of the thin shell rigid object (if any) that separates particle and its surrounding grid cell
-            self.cpic_flag = ti.field(gs.ti_int, shape=(self.mpm_solver.n_particles, 3, 3, 3))
+            self.cpic_flag = ti.field(gs.ti_int, shape=(self.mpm_solver.n_particles, 3, 3, 3, self.mpm_solver._B))
             self.mpm_rigid_normal = ti.Vector.field(
-                3, dtype=gs.ti_float, shape=(self.mpm_solver.n_particles, self.rigid_solver.n_geoms_)
+                3,
+                dtype=gs.ti_float,
+                shape=(self.mpm_solver.n_particles, self.rigid_solver.n_geoms_, self.mpm_solver._B),
             )
 
         if self._rigid_sph:
             self.sph_rigid_normal = ti.Vector.field(
-                3, dtype=gs.ti_float, shape=(self.sph_solver.n_particles, self.rigid_solver.n_geoms_)
+                3,
+                dtype=gs.ti_float,
+                shape=(self.sph_solver.n_particles, self.rigid_solver.n_geoms_, self.sph_solver._B),
             )
             self.sph_rigid_normal_reordered = ti.Vector.field(
-                3, dtype=gs.ti_float, shape=(self.sph_solver.n_particles, self.rigid_solver.n_geoms_)
+                3,
+                dtype=gs.ti_float,
+                shape=(self.sph_solver.n_particles, self.rigid_solver.n_geoms_, self.sph_solver._B),
             )
 
         if self._rigid_pbd:
-            self.pbd_rigid_normal = ti.Vector.field(
-                3, dtype=gs.ti_float, shape=(self.pbd_solver.n_particles, self.rigid_solver.n_geoms)
-            )
+            # self.pbd_rigid_normal = ti.Vector.field(
```

Comment:
Do not merge comment in the codebase without good motivation clearly stated by a comment.